Mikhails/homework template

Mikhail-SHA-256 opened the pull request:
This is just template implementation part of homework

Mikhail-SHA-256 committed:
Introduce VectorCommon and Vector2d as specification of common vector

Mikhail-SHA-256 committed:
Implement Vector3d as specification of common vector

maciekgajewski commented on course6/toyapp_mikhail/displaylib/vector2d.h:
```diff
@@ -1,26 +1,9 @@
-			r.y += v.y;
-			return r;
-		}
+using Vector2D = VectorCommon<double, 2>;
```
I would say it's statistical overkill to have a separate source file for one line

Mikhail-SHA-256 replied:
Fixed. Left just one file with template definitions

maciekgajewski commented on course6/toyapp_mikhail/displaylib/vector_common.h:
```diff
@@ -0,0 +1,57 @@
+template <class TValue, int TDimentions>
+class VectorCommon
+{
+	// No constructor for vector without specification
```
Simply provide no definition for primary template at all, then only specializations will be allowed. 
But in this case I was expecting an actual primary template that works with any size

maciekgajewski commented on course6/toyapp_mikhail/displaylib/vector_common.h:
```diff
@@ -0,0 +1,57 @@
+};
+
+// Specification for 2 dimentional vector
+template <class TValue>
```
no no no, this is not what I had in mind. The goal of this template should be to use vector of any size/type

Mikhail-SHA-256 replied:
Changed.
But I still want to use nice decorations with x,y and x,y,z to not change client code.
Using v.x, v.y in client code is cleaner than v[0], v[1].
For decorations I had to use inheritance. is it ok?

maciekgajewski replied:
Forget about ```v.x```, ```v.y```. C++ doesn't have the support for properties.
The canonical way is to use accessor methods: getters and setters.

You may implement ```v.x()```, ```v.y()``` in an inherited class, and they can be an alternative api to ```v[0]```.

You will have to change your client code one way or the other. Consider this a punishment for using public fields :)

Mikhail-SHA-256 replied:
Ok

maciekgajewski reviewed (changes requested):
This is not what I expected. See other's solutions

Mikhail-SHA-256 committed:
Homework Template: Fix code review issues.

Move operators into main template.
Use decorator classes to keep existing API of Vector2d, Vector3D.

Mikhail-SHA-256 committed:
Homework Template: Fix code review issues.

Use methods which returns references instead of reference data fields.

Mikhail-SHA-256 committed:
#nocode Automatic reformat indent in all files

no functional changes

Mikhail-SHA-256 committed:
#nocode Fix code review issues. Naming, code move h->cc

Move method implementation code from h to cc. Fix naming issue for member data.

Mikhail-SHA-256 commented:
Add more fixes based on feedback

Separate commit for reformatting

maciekgajewski commented on course6/toyapp_mikhail/displaylib/vector_common.h:
```diff
@@ -0,0 +1,100 @@
+class VectorCommon
+{
+private:
+	using TSelf = VectorCommon<TValue, TDimentions>;
```
You can just use ```VectorCommon``` inside of the class' definition.

maciekgajewski commented on course6/toyapp_mikhail/displaylib/vector_common.h:
```diff
@@ -0,0 +1,100 @@
+
+namespace Display {
+
+template <class TValue, size_t TDimentions, class TDecoratedType = TValue>
```
the "T" in "TDimensions" doesn't styand for "type"? right?
The popular convention is to use ALL_CAPS for value template arguments. Just like for constants; they are compile-time constants after all.

maciekgajewski replied:
Why is "TDecoratedType"? It's not imemdiately obvious for mew why would you need that in a vector, which means it needs eithjer better name, or at least explanation in a comment.

maciekgajewski replied:
Ah! Decorated type is the implementation! You are doing CRTP! Then using TValue as default makes no sense, and TSelf is useless.

Mikhail-SHA-256 replied:
namespace Display::Detail

maciekgajewski commented on course6/toyapp_mikhail/displaylib/vector_common.h:
```diff
@@ -0,0 +1,100 @@
+	{}
+};
+
+using Vector2D = Vector2DimDecorator<double>;
```
Decorator/Decorated is not a canonical naming convention.
```Vector2Dim``` would be a better name, and ```VectorCommon``` should rather be ```VectorBase```

maciekgajewski reviewed (changes requested):
Good use of CRTP!
But the naming and some details make it confusing.

Mikhail-SHA-256 committed:
Fixed core review issues

Mikhail-SHA-256 committed:
Fix compilation errors

Mikhail-SHA-256 committed:
Rename TDecoratedType parameter

Mikhail-SHA-256 committed:
Remove explicit constructors of Vector2, Vectro3 in favor of implicit copy constructor

maciekgajewski reviewed (approved):
Good!
